### PR TITLE
Implemented CA1725 + Fixer

### LINKED ISSUE #457: Port FxCop rule CA1725: ParameterNamesShouldMatchBaseDeclaration
**Title:** Parameter names should match base declaration

**Description:**

Consistent naming of parameters in an override hierarchy increases the usability of the method overrides. A parameter name in a derived method that differs from the name in the base declaration can cause confusion about whether the method is an override of the base method or a new overload of the method.

**Dependency:** None

**Notes:**

See if StyleCop has implemented it

### PR DESCRIPTION
**Pull request by vweijsters**
Fixes #457 

The implementation is based on DotNetAnalyzers/StyleCopAnalyzers#2015

### CODE REVIEW

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs [hunk 1]
```diff
@@ -29,10 +34,113 @@ public abstract class ParameterNamesShouldMatchBaseDeclarationAnalyzer : Diagnos
                                                                              helpLinkUri: null,     // TODO: add MSDN url
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
+        private static readonly Action<SymbolAnalysisContext> AnalyzeMethodSymbolAction = AnalyzeMethodSymbol;
```
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs]: Is this field required? Can't you directly pass `AnalyzeMethodSymbol` to `RegisterSymbolAction`?

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs [hunk 2]
```diff
@@ -29,10 +34,113 @@ public abstract class ParameterNamesShouldMatchBaseDeclarationAnalyzer : Diagnos
...
+
+                    if (currentParameter.Name == originalParameter.Name)
+                    {
+                        currentMatchScore++;
+                    }
+                }
+
+                if (currentMatchScore > bestMatchScore)
+                {
+                    bestMatch = originalDefinition;
```
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs]: I like your approach of suggesting the best match by count of matching parameter names. FXCop actually searches through the list of original definitions and reports an issue for each definition against which the match fails - that seems like lot of clutter.

Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs]: Actually, my bad. Even FXCop reports violation only on the first non-matching definition. However, it just picks the first one.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs [hunk 3]
```diff
@@ -29,10 +34,113 @@ public abstract class ParameterNamesShouldMatchBaseDeclarationAnalyzer : Diagnos
...
+                    bestMatchScore = currentMatchScore;
+                }
+            }
+
```
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs]: We should bail out here if `bestMatchScore == methodSymbol.Parameters.Length`

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs [hunk 4]
```diff
@@ -4,15 +4,20 @@
...
     /// CA1725: Parameter names should match base declaration
     /// </summary>
-    public abstract class ParameterNamesShouldMatchBaseDeclarationAnalyzer : DiagnosticAnalyzer
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class ParameterNamesShouldMatchBaseDeclarationAnalyzer : DiagnosticAnalyzer
     {
         internal const string RuleId = "CA1725";
+        internal const string NewNamePropertyName = "NewName";
 
         private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.ParameterNamesShouldMatchBaseDeclarationTitle), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
```
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ParameterNamesShouldMatchBaseDeclaration.cs]: This is not your fault, but the message resource string `ParameterNamesShouldMatchBaseDeclarationMessage` is not correct - it should be parameterized with the actual method name. As we are trying to retain as much compat with FXCop as possible, you should change it to:

```
In member {0}, change parameter name {1} to {2} in order to match the identifier as it has been declared in {3}.
```
where {0} = violating member, {1} = violating parameter name, {2} = base parameter name, {3} = base member

**Comment by vweijsters** [comment]
Updated and rebased after comments

**Comment by mavasani** [comment]
👍 Thanks for the contribution @vweijsters 
I'll merge this in now..

**Pull request merged**